pkg/metrics/instance/configstore/kv/consul/client.go

```diff
@@ -302,7 +305,13 @@ func (c *Client) WatchPrefix(ctx context.Context, prefix string, f func(string,
 			continue
 		}
 
+		newKeys := make(map[string]struct{}, len(kvps))
+
 		for _, kvp := range kvps {
+			// Keep track of all returned keys. This *must* be done before the check
+			// below otherwise an unchanged key will be treated as deleted.
+			newKeys[kvp.Key] = struct{}{}
```

Comment:
Do we also need to keep track of the keys in List? We might be good if the watch always executes before the list but relying on that feels not great.

Reply:
Hmm, not without also changing etcd to do the same. I think treating WatchPrefix as "you will only get CRUD events from the start of the watch" makes sense, though.